subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,303 @@
+			    CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+			memcpy(hc_uart->rx_ctx->buf + hc_uart->rx_ctx->len,
+			       evt->data.rx.buf + evt->data.rx.offset, evt->data.rx.len);
+		} else if ((hc_uart->rx_ctx->len + evt->data.rx.len) >
+			   CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+			hc_uart->state = USART_HOST_CMD_RX_BAD;
+			return;
+		}
+
+		hc_uart->rx_ctx->len += evt->data.rx.len;
+
+		if (hc_uart->rx_ctx->len >= sizeof(struct ec_host_cmd_request_header)) {
+			/* Buffer has request header. Check header and get data_len */
+			hc_uart->expected_len = request_expected_size(
+				(struct ec_host_cmd_request_header *)hc_uart->rx_ctx->buf);
+
+			if (hc_uart->expected_len == 0 ||
+			    hc_uart->expected_len > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER) {
+				/* Invalid expected size, set the state and wait for timeout */
+				hc_uart->state = USART_HOST_CMD_RX_BAD;
+			} else if (hc_uart->rx_ctx->len == hc_uart->expected_len) {
+				/* Don't wait for overrun, because it is already done
+				 * in a UART driver.
+				 */
+				(void)k_work_cancel_delayable(&hc_uart->timeout_work);
+
+				/* TODO: Workaround - reprogram DMA not to lose data at the end of
+				 * buffer. Remove it once the circular buffer is supported.
+				 */
+				uart_rx_disable(hc_uart->uart_dev);
+				uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+					       CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER,
+					       OVERRUN_TIMEOUT);
+
+				/* If no data more in request, packet is complete. Start processing
+				 */
+				hc_uart->state = USART_HOST_CMD_PROCESSING;
+
+				k_sem_give(&hc_uart->rx_ctx->handler_owns);
+			} else if (hc_uart->rx_ctx->len > hc_uart->expected_len) {
+				/* Overrun error, set the state and wait for timeout */
+				hc_uart->state = USART_HOST_CMD_RX_OVERRUN;
+			}
+		}
+		break;
+	case UART_RX_BUF_REQUEST:
+		/* Do not provide the second buffer, because we reload DMA after every packet.
+		 * It should be changed once the circular buffer is supported.
+		 */
+		break;
+	case UART_TX_DONE:
+		(void)k_work_cancel_delayable(&hc_uart->timeout_work);
+		hc_uart->state = USART_HOST_CMD_READY_TO_RX;
+
+		break;
+	default:
+		break;
+	}
+}
+
+static int ec_host_cmd_uart_init(const struct ec_host_cmd_backend *backend,
+				 struct ec_host_cmd_rx_ctx *rx_ctx, struct ec_host_cmd_tx_buf *tx)
+{
+	int ret;
+	struct ec_host_cmd_uart_ctx *hc_uart = (struct ec_host_cmd_uart_ctx *)backend->ctx;
+
+	hc_uart->state = USART_HOST_CMD_STATE_DISABLED;
+
+	if (!device_is_ready(hc_uart->uart_dev)) {
+		return -ENODEV;
+	}
+
+	/* UART backend needs rx and tx buffers provided by the handler */
+	if (!rx_ctx->buf || !tx->buf) {
+		return -EIO;
+	}
+
+	hc_uart->rx_ctx = rx_ctx;
+	hc_uart->tx = tx;
+
+	k_work_init_delayable(&hc_uart->timeout_work, rx_timeout);
+	uart_callback_set(hc_uart->uart_dev, uart_callback, (void *)hc_uart);
+	ret = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER, OVERRUN_TIMEOUT);
```

Comment:
Should have `CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER` been `CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER_SIZE`?